streams/src/test/java/org/apache/kafka/streams/processor/internals/TransactionManagerTest.java

```diff
@@ -0,0 +1,638 @@
+import org.apache.kafka.common.errors.TimeoutException;
+import org.apache.kafka.common.errors.UnknownProducerIdException;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.apache.kafka.streams.errors.TaskMigratedException;
+import org.apache.kafka.streams.processor.TaskId;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.kafka.common.utils.Utils.mkEntry;
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.easymock.EasyMock.expect;
+import static org.easymock.EasyMock.expectLastCall;
+import static org.easymock.EasyMock.mock;
+import static org.easymock.EasyMock.replay;
+import static org.easymock.EasyMock.verify;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+public class TransactionManagerTest {
+
+    private final LogContext logContext = new LogContext("test ");
+    private final String topic = "topic";
+    private final Cluster cluster = new Cluster(
+        "cluster",
+        Collections.singletonList(Node.noNode()),
+        Collections.singletonList(new PartitionInfo(topic, 0, Node.noNode(), new Node[0], new Node[0])),
+        Collections.emptySet(),
+        Collections.emptySet()
+    );
+    private final TaskId taskId = new TaskId(0, 0);
+    private final ByteArraySerializer byteArraySerializer = new ByteArraySerializer();
+    private final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata = mkMap(
+        mkEntry(new TopicPartition(topic, 0), new OffsetAndMetadata(0L, null))
+    );
+
+    private final MockProducer<byte[], byte[]> mockProducer = new MockProducer<>(
+        cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer);
+    private final TransactionManager transactionManager = new TransactionManager(logContext, mockProducer);
+
+    private final MockProducer<byte[], byte[]> eosMockProducer = new MockProducer<>(
+        cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer);
+    private final TransactionManager eosTransactionManager = new TransactionManager(logContext, eosMockProducer, "appId", taskId);
+
+    private final ProducerRecord<byte[], byte[]> record =
+        new ProducerRecord<>(topic, 0, 0L, new byte[0], new byte[0], new RecordHeaders());
+
+    @Before
+    public void before() {
+        eosTransactionManager.initTransaction();
+    }
+
+    @Test
+    public void shouldFailIfProducerIsNull() {
+        {
```

Comment:
Brackets looks unnecessary here.

Reply:
I reuse the variable name `thrown` twice and want to make it `final` -- hence the braces